### PR TITLE
improve `F.group_normalization` via cuDNN call

### PR DESCRIPTION
**Pull request by grafi-tt**
Memory consumption reduced and speed slightly improved.

* cuDNN's batch normalization routine is directly called to simplify computation graph.
* `x_hat = (x - mean) / std` is re-computed at backward phase to reduce memory consumption.
* Some `ElementwiseKernel` and `ReductionKernel` are used to reduce kernel calls.

Note that if we cache `x_hat`, speed can be further improved. But I think memory consumption is more crucial, because Group Normalization is particularly useful when batch size can't be increased due to memory limit.

Regarding speed, time to train ResNet50 with 224 x 224 input images becomes 94.5%. Validation was done at each epoch, and train/valid split ratio was 8/2. Batch size was 32. The benchmark environment is as follows.

* CPU: Ryzen7 2700X
* GPU: ZOTAC GTX 1080 Ti Amp Edition
* DRAM: DDR4-3200 8GiB x 2 dual channel
* SSD: WD Black NVMe 1TB
* OS: Linux 64bit
* CUDA version: 10.0.130
* cuDNN version: 7.4.1.5

Add to that, as Group Normalization is a generalization of Layer Normalization and Instance Normalization, current `F.layer_normalization` implementation can be replaced in future.

### CODE REVIEW

**Comment by hvy** [comment]
Thanks for this PR and sorry for the delayed response. Could you post the memory usage improvement here as well, in addition to the speed improvements?

File: chainer/functions/normalization/group_normalization.py [hunk 1]
```diff
@@ -1,10 +1,348 @@
...
+        gx = gx.reshape(orig_shape)
+        return gx, ggamma, gbeta
+
+
+class _GradHelper(function_node.FunctionNode):
+
+    def forward(self, inputs):
+        self.retain_inputs((0, 1, 2))
+        x_hat, gamma, gy = inputs
+        xp = backend.get_array_module(x_hat)
```
Comment (hvy) [inline chainer/functions/normalization/group_normalization.py]: A detail but let's declare `xp` closer to where it's used. Also, it could be removed since it's only used once.

File: chainer/functions/normalization/group_normalization.py [hunk 2]
```diff
@@ -1,10 +1,348 @@
...
+        gy = gy.reshape((batch_size, channels, -1))
+        ggx_hat = ggx_hat.reshape((batch_size, channels, -1))
+
+        gx_hat2 = gggamma[:, None] * gy
+        ggamma2 = chainer.functions.sum(ggx_hat * gy, axis=(0, 2))
+        ggy = (ggx_hat * gamma[:, None] + gggamma[:, None] * x_hat +
+               ggbeta[:, None])
+
+        gx_hat2 = chainer.functions.reshape(gx_hat2, reduced_shape)
+        ggy = chainer.functions.reshape(ggy, orig_shape)
```
Comment (hvy) [inline chainer/functions/normalization/group_normalization.py]: Just curious but how come you're using the free function `reshape` instead of the method here (you're mostly using the method in other places)?

File: chainer/functions/normalization/group_normalization.py [hunk 3]
```diff
@@ -1,10 +1,348 @@
...
+
+        self.groups = groups
+        self.eps = eps
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 3)
+        x_type, gamma_type, beta_type = in_types
+        type_check.expect(
+            x_type.dtype.kind == 'f',
+            x_type.ndim >= 2,
```
[the remark this answers is not in the document]
Comment (toslunar) [reply chainer/functions/normalization/group_normalization.py]: The existing check is too strict.  My suggestion: grafi-tt#2

**Comment by grafi-tt** [comment]
@toslunar Thank you for your PR!

@hvy Thank you for your review, and sorry for late reply. I reflected your suggestions. Could you make review again?

File: chainer/functions/normalization/group_normalization.py [hunk 4]
```diff
@@ -1,10 +1,360 @@
...
+
+        batch_size, channels = gy.shape[:2]
+        gy = gy.reshape((batch_size, channels, -1))
+
+        reduced_shape = x_hat.shape
+        x_hat = x_hat.reshape((batch_size, channels, -1))
+
+        gx_hat = gy * gamma[:, None]
+        gbeta = gy.sum(axis=(0, 2))
+        if backend.get_array_module(x_hat) is numpy:
```
Comment (toslunar) [inline chainer/functions/normalization/group_normalization.py]: ditto

File: chainer/functions/normalization/group_normalization.py [hunk 5]
```diff
@@ -1,10 +1,360 @@
...
+        gx_hat, ggamma, gbeta = _GradHelper().apply((x_hat, gamma, gy))
+        gx, = _XHatGrad(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma, x_hat.array).apply((x, gx_hat))
+
+        gx = gx.reshape(orig_shape)
+        return gx, ggamma, gbeta
+
+
+class _GradHelper(function_node.FunctionNode):
```
Comment (toslunar) [inline chainer/functions/normalization/group_normalization.py]: I'd mention the function node computes grad of `y = gamma * x_hat + beta`.  Maybe it's sufficient to rename to a more expressive name like `_ScaleShiftGrad`.

File: chainer/functions/normalization/group_normalization.py [hunk 6]
```diff
@@ -1,10 +1,360 @@
...
+        self.dummy_gamma = dummy_gamma
+
+    def forward(self, inputs):
+        self.retain_inputs((0,))
+        _, y = inputs
+        z = self.inv_std[:, None] * y
+        self.retain_outputs((0,))
+        return z,
+
+    def backward(self, indexes, grad_outputs):
```
Comment (toslunar) [inline chainer/functions/normalization/group_normalization.py]: It seems the method is not tested because it's used for >=3rd order grad.

**Comment by hvy** [comment]
Thanks for the fixes, they look good to me. Could you also take a look at the remaining ones? Also, could you post some benchmarks on the memory consumption improvements (as requested in a comment above)?

File: chainer/functions/normalization/group_normalization.py [hunk 7]
```diff
@@ -1,10 +1,360 @@
...
+        ggx, = grad_outputs
+
+        x_hat, = _XHat(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma).apply((x,))
+
+        ret = []
+
+        if 0 in indexes:
+            # gx = inv_std * gx_std
```
Comment (crcrpar) [inline chainer/functions/normalization/group_normalization.py]: I think these commented out lines can be removed.

Comment (grafi-tt) [reply chainer/functions/normalization/group_normalization.py]: Those lines are *not* commented out source codes. They are here to sketch the derivation of double backprop formula.

I'd like to make the comments less confusing, but I don't have an idea...

Comment (crcrpar) [reply chainer/functions/normalization/group_normalization.py]: apology.

How about adding a single comment that clarifies below lines are not comment, but the sketch of double backprop formula?
For me, just adding some comments at the beginning and end of the sketch is enough.

**Comment by grafi-tt** [comment]
@toslunar @hvy @crcrpar Thank you very much. I fixed the problems commented.

* rebase on v6.0.0b3
* compare `xp` to `cuda.cupy`
* rename `_GradHelper` to `_ScaleBiadGrad`
* fix `_MulInvStd` backward computation (sign flipped)
* make comments about `gx2` less confusing
* simplify tests using `FunctionTestCase`
* add test for `_MulInvStd`
* test with large `eps` value

I checked runtime memory consumption of ResNet50 by `nvidia-smi`. It reduced from 7473MiB to 3977MiB. Though this is a rough measure, I think it is enough to tell the improvement.

**Comment by toslunar** [comment]
Great!

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
cupy/cupy#2060 broke this PR.  Let me wait for cupy/cupy#2072.

**Comment by takagi** [comment]
@grafi-tt Would you merge the current master to your branch? The master is based on the change of cupy/cupy#2060, so your PR should pass the CI after merging it.

**Comment by grafi-tt** [comment]
@takagi Thanks. I merged the master branch #a8f9cb5. But If I understand correctly it doesn't fix the issue, because the GroupNormalization implementation directly calls `cudnn.batch_normalization_forward_training`. Should I modify the code according to #2060?

**Comment by toslunar** [comment]
As #6323 fixed a fp16 issue, the PR needs the same fix (to stop using `F.batch_normalization`).

**Comment by takagi** [comment]
Ah, I read `GroupNormalization ` of the master, where `cudnn.batch_normalization_forward_training` is not called directly. Yes, it would be great if you modified your code as well.

**Comment by grafi-tt** [comment]
@takagi Thank you. I added the modification like #6323.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Sorry for the late response, and thanks a lot for working on this long running PR. LGTM!
